config/v1/types_image_digest_mirror_set.go

```diff
@@ -0,0 +1,113 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ImageDigestMirrorSetList lists the items in the ImageDigestMirrorSet CRD.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type ImageDigestMirrorSetList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ImageDigestMirrorSet `json:"items"`
+}
+
+// ImageDigestMirrors holds cluster-wide information about how to handle mirrors in the registries config.
+type ImageDigestMirrors struct {
+	// source matches the repository that users refer to, e.g. in image pull specifications.
+	// "source" uses one of the following formats:
+	// host[:port]
+	// host[:port]/namespace[/namespace…]
+	// host[:port]/namespace[/namespace…]/repo
+	// [*.]host
+	// for more information about the format, see the document about the location field:
+	// https://github.com/containers/image/blob/main/docs/containers-registries.conf.5.md#choosing-a-registry-toml-table
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^\*(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+$|^((?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?)(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?$`
+	Source string `json:"source"`
+	// mirrors is zero or more locations that may also contain the same images.
+	// Images can be pulled from these mirrors only if they are referenced by their digests.
+	// The mirrored location is obtained by replacing the part of the input reference that
+	// matches source by the mirrors entry, e.g. for registry.redhat.io/product/repo reference,
+	// a (source, mirror) pair *.redhat.io, mirror.local/redhat causes a mirror.local/redhat/product/repo
+	// repository to be used.
+	// If the "mirrors" is not specified, the image will continue to be pulled from the specified
+	// repository in the pull spec. No mirror will be configured.
+	// The order of mirrors in this list is treated as the user's desired priority, while source
+	// is by default considered lower priority than all mirrors. Other cluster configuration,
+	// including (but not limited to) other imageDigestMirrors objects,
+	// may impact the exact order mirrors are contacted in, or some mirrors may be contacted
+	// in parallel, so this should be considered a preference rather than a guarantee of ordering.
+	// "mirrors" uses one of the following formats:
+	// host[:port]
+	// host[:port]/namespace[/namespace…]
+	// host[:port]/namespace[/namespace…]/repo
+	// for more information about the format, see the document about the location field:
+	// https://github.com/containers/image/blob/main/docs/containers-registries.conf.5.md#choosing-a-registry-toml-table
+	// +optional
+	// +listType=set
+	Mirrors []Mirrors `json:"mirrors,omitempty"`
```

Comment:
Is my registry authentication presented to these mirrors?  Given multiple possible tokens, are tokens tried against each mirror first or is a single token tried against all mirrors first?

Reply:
tokens tried against each mirror. The order of mirrors in this mirrors list is the desired order. The mirrors will be tried in this order, if you have multiple tokens including the token for the current pulling mirror, the token will be used to access the mirror.

Reply:
All of these mirrors configurations are invisible to Kubelet, and the CRI (the Kubelet / CRI-O API) does not currently allow passing credentials for multiple registries.

So, what happens is that the Kubelet reads the Pod’s pull secret, if any, and extracts the credentials only matching _source_, and provides them in the CRI API. If the Pod’s pull secret contains any credentials for mirrors, those are completely ignored. CRI-O uses the CRI-provided credentials to contact the source only; for mirrors, it uses credentials available in the node-global `config.json` (in OpenShift, that’s the global cluster pull secret).

---

(IIRC, if the Pod’s pull secret(s) contain multiple sets of credentials for the same registry, Kubelet tries the pull operation, with each credential. So, in that case, we would try all of the credentials for the source registry; but the mirrors, if any, would all use the global cluster pull secret, repeatedly on each of those retries.)

---

(The inability of the user to provide mirror credentials via a pod’s pull secret is a known limitation of the mirroring mechanism, and nothing changes about that with this enhancement / these API changes.)